specification/theory_operations.adoc

```diff
@@ -1,2 +1,277 @@
+
 == Theory of Operations
+
+The CoVE-IO specification extends the CoVE host and guest SBI extensions to
+allow TVMs to establish trust with TEE-IO devices, and then use and interact
+with those devices. Untrusted domain components are responsible for assigning
+TEE-IO devices to TVMs, and also for supporting the TVM acceptance or rejection
+of the assigned devices. The TSM, on the other hand, establishes and maintains
+the secure physical and logical links between TVMs and their assigned devices.
+
+The following sections describe the functionality of the TSM provided CoVE-IO
+extensions to support trusted I/O on CoVE enabled platforms.
+
+=== Platform Initialization
+
+==== IOMMU Registration and Setup
+
+The TSM relies on the availability of at least one IOMMU instance exclusvely
+associated with the TSM trusted domain. Those IOMMUs allow the TSM to perform
+address translations and protection from DMA and interrupts originating from
+assigned TDIs. Only the TSM can access and program he trusted domain IOMMU
+instances.
+
+Trusted domain assigned IOMMUs can generate MSIs in order to signal the TSM
+about command completions, transaction faults or device page requests. Those
+MSIs target system physical memory, which is owned by the untrusted domain
+manager, e.g. the host VMM. As a consequence, it is the untrusted domain
+responsibility to reserve the MSI addresses and then rely on the TSM to program
+the IOMMUs with those reserved addresses. This IOMMU registration process is
+driven by the untrusted domain manager for all IOMMUs that participate in TEE-IO
+and gies as follows:
+
+1. The TSM is loaded and provisioned with a CoVE-IO manifest.
+2. The untrusted domain manager (e.g. the host VMM) enumerates all platform
+   IOMMUs as PCIe devices.
+3. The untrusted IOMMU driver is loaded and probed.
+4. The untrusted IOMMU driver allocates MSI vectors for the trusted domain
+   IOMMUs. Those vectors must point to the untrusted IOMMU driver as the TSM can
+   not handle external interrupts.
+5. The untrusted domain manager registers the IOMMUs with the TSM by calling
+   the `sbi_covh_register_iommu()` COVH function. The TSM gets the allocated MSI
+   vectors and configures the trusted IOMMU `s_msi_cfg_table` register
+   accordingly.
+6. When a trusted IOMMU sends an MSI, the untrusted IOMMU driver handles it and
+   notifies the TSM about a pending trusted IOMMU MSI by calling the
+   `sbi_covh_notify_iommu_msi()` COVH function.
+7. The TSM verifies that there is a pending MSI by reading the IOMMU `s_ipsr`
+   register, and handles the interrupts as needed.
+
+The untrusted driver can trick the TSM by either invoking the
+`sbi_covh_notify_iommu_msi()` COVH function while there are no pending MSI from
+the trusted IOMMU or not invoking it when there actually is one such pending
+interrupt. In the former case, the TSM can verify that there really is a pending
+MSI by checking the trusted IOMMU status registers. The latter case could cause
```

Comment:
have we introduced what we mean by "trusted IOMMU status registers"?

Reply:
We have not, I'll rephrase.